content/blog/2020/how-selenium-works-2-navigation.md

```diff
@@ -0,0 +1,70 @@
+
+One of these pains is certificates. Companies are cheap and will do self-signed certificates and other monstrosities. Especially in the early days of Selenium when there wasn't services like [Let's Encrypt](https://letsencrypt.org/). And even now, most Developers and QA teams rarely have access to change configurations on their test environments or on their CI servers. We needed to come up with a way to circumvent the certificates. (_This is one of the first reasons why Selenium is seen as a security risk_)
+
+So each of the browsers will allow the poorly configured certificates through when we automate. It's because if we didn't a lot of testers/developers would never be able to test their sites. 
+
+Now... that we go round the first problem we need to move on to getting the page to load.
+
+## Loading
+
+Once we have got round the certificates we get the page to load. Luckily we don't need to do anything more than the equivalent of 
+
+{{< highlight javascript >}}
+location = "https://www.theautomatedtester.co.uk"; 
+// or
+window.location.href = "https://www.theautomatedtester.co.uk";
+{{< /highlight >}}
+
+## Done... 
+
+When Selenium has "finished" a command it will return. So, we just need to wait for the page to be finished loading. To be honest, what does _"finished loading"_ even mean.
+
+In the browser it will fire a number of different events. We will know when the page has been shown and then tell you what it's [`readyState`](https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState). Selenium will check all of that and it will also wait for the [`DOMContentLoaded`](https://developer.mozilla.org/en-US/docs/Web/API/Window/DOMContentLoaded_event).
+
+Once we have done those checks, and you can manipulate how we look at those events, if you want it to load quicker, using the [Page Load Strategies](https://w3c.github.io/webdriver/#dfn-page-loading-strategy). This is more of a power user feature so I wouldn't worry about them now.
+
+## What about JavaScript Frameworks and navigation
+
+This is where there is "fun". A lot of frameworks will do a lot of loading after their initial load on to the page. If you have ever worked on a single page app, or have ever used a single page app, you will have seen a lot of items render as the data comes in. Unfortunately it does mean that you can't just rely on the navigation command returning. You will have to add a `WebDriverWait` command to your code, as shown below, to make sure your test is in the right state before going off to do what it needs to.
+
+{{< highlight python >}}
+from selenium import webdriver
+from selenium.webdriver.support.wait import WebDriverWait
+
+driver = webdriver.Firefox()
+driver.get("https://www.theautomatedtester.co.uk")
+element = WebDriverWait(driver, 10).until(lambda x: x.find_element_by_id(“someId”))
+{{< /highlight >}}
+
+## Further Reading
```

Comment:
How do implicit waits play into the page loading strategies? You mention in passing that Selenium will wait for a command to be "finished". What does that mean?